File: android_app/src/main/java/app/time_to/timeto/ui/FoldersSettingsSheet.kt
package app.time_to.timeto.ui

import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Icon
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import app.time_to.timeto.R
import app.time_to.timeto.rememberVM
import app.time_to.timeto.setTrue
import timeto.shared.vm.FoldersSettingsVM

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun FoldersSettingsSheet(
    isPresented: MutableState<Boolean>,
) {
    TimetoSheet(isPresented = isPresented) {

        val (vm, state) = rememberVM { FoldersSettingsVM() }

        Column(
            modifier = Modifier
                .fillMaxHeight()
                .background(c.bgFormSheet)
        ) {

            val scrollState = rememberLazyListState()

            SheetHeaderView(
                onCancel = { isPresented.value = false },
                title = state.headerTitle,
                doneText = state.headerDoneText,
                isDoneEnabled = true,
                scrollToHeader = 0, // todo
            ) {
                isPresented.value = false
            }

            LazyColumn(
                state = scrollState,
                modifier = Modifier
                    .padding(bottom = 20.dp)
                    .navigationBarsPadding()
                    .imePadding(),
                contentPadding = PaddingValues(top = 20.dp, bottom = 20.dp)
            ) {

                val folders = state.folders
                items(
                    items = folders,
                    key = { folder -> folder.id }
                ) { folder ->

                    val isFirst = folders.first() == folder

                    val isEditFolderPresented = remember { mutableStateOf(false) }
                    FolderFormSheet(
                        isPresented = isEditFolderPresented,
                        folder = folder,
                    )

                    MyListView__ItemView(
                        isFirst = isFirst,
                        isLast = folders.last() == folder,
                        withTopDivider = !isFirst,
                        modifier = Modifier.animateItemPlacement(),
                    ) {
                        MyListView__ItemView__ButtonView(
                            text = folder.name,
                            rightView = {
                                Icon(
                                    painterResource(id = R.drawable.ic_round_arrow_upward_24),
                                    "Up",
                                    tint = c.blue,
                                    modifier = Modifier
                                        .padding(end = 10.dp)
                                        .size(26.dp)
                                        .clip(RoundedCornerShape(99.dp))
                                        .clickable(!isFirst) {
                                            vm.sortUp(folder)
                                        }
                                        .padding(2.dp)
                                )
                            }
                        ) {
                            isEditFolderPresented.setTrue()
                        }
                    }
                }

                item {

                    MyListView__Padding__SectionSection()

                    val isAddFolderPresented = remember { mutableStateOf(false) }
                    FolderFormSheet(
                        isPresented = isAddFolderPresented,
                        folder = null,
                    )

                    MyListView__ItemView(
                        isFirst = true,
                        isLast = true,
                    ) {
                        MyListView__ItemView__ButtonView(
                            text = "New Folder",
                        ) {
                            isAddFolderPresented.setTrue()
                        }
                    }
                }

                ////

                val tmrwButtonUI = state.tmrwButtonUI
                if (tmrwButtonUI != null) {
                    item {

                        MyListView__Padding__SectionSection()

                        MyListView__ItemView(
                            isFirst = true,
                            isLast = true,
                        ) {
                            MyListView__ItemView__ButtonView(
                                text = tmrwButtonUI.text,
                            ) {
                                tmrwButtonUI.add()
                            }
                        }
                    }
                }
            }
        }
    }
}

File: ios_app/iphone/FoldersSettingsSheet.swift
import SwiftUI
import shared

struct FoldersSettingsSheet: View {

    @Binding var isPresented: Bool

    @State private var vm = FoldersSettingsVm()
    @State private var sheetHeaderScroll = 0

    @State private var isAddSheetPresented = false

    var body: some View {

        VMView(vm: vm, stack: .VStack(spacing: 0)) { state in

            SheetHeaderView(
                    onCancel: { isPresented.toggle() },
                    title: state.headerTitle,
                    doneText: nil,
                    isDoneEnabled: false,
                    scrollToHeader: sheetHeaderScroll,
                    cancelText: "Back"
            ) {}

            ScrollViewWithVListener(showsIndicators: false, vScroll: $sheetHeaderScroll) {

                VStack {

                    Padding(vertical: 20)

                    let folders = state.folders
                    ForEach(folders, id: \.id) { folder in
                        let isFirst = folders.first == folder
                        MyListView__ItemView(
                                isFirst: isFirst,
                                isLast: folders.last == folder,
                                withTopDivider: !isFirst
                        ) {
                            FoldersSettingsSheet__FolderItem(
                                    vm: vm,
                                    folder: folder
                            )
                        }
                    }

                    MyListView__Padding__SectionSection()

                    MyListView__ItemView(
                            isFirst: true,
                            isLast: true
                    ) {
                        MyListView__ItemView__ButtonView(
                                text: "New Folder"
                        ) {
                            isAddSheetPresented = true
                        }
                                .sheetEnv(isPresented: $isAddSheetPresented) {
                                    FolderFormSheet(
                                            isPresented: $isAddSheetPresented,
                                            folder: nil
                                    )
                                }
                    }

                    if let tmrwButtonUi = state.tmrwButtonUi {

                        MyListView__Padding__SectionSection()

                        MyListView__ItemView(
                                isFirst: true,
                                isLast: true
                        ) {
                            MyListView__ItemView__ButtonView(
                                    text: tmrwButtonUi.text
                            ) {
                                tmrwButtonUi.add()
                            }
                        }
                    }
                }
            }
        }
                .background(c.sheetBg)
    }
}

private struct FoldersSettingsSheet__FolderItem: View {

    let vm: FoldersSettingsVm
    let folder: TaskFolderDb

    @State private var isEditPresented = false

    var body: some View {

        MyListView__ItemView__ButtonView(
                text: folder.name,
                rightView: AnyView(
                        Button(
                                action: {
                                    vm.sortUp(folder: folder)
                                },
                                label: {
                                    Image(systemName: "arrow.up")
                                            .font(.system(size: 14))
                                            .foregroundColor(.blue)
                                }
                        )
                                .buttonStyle(.plain)
                                .padding(.trailing, H_PADDING)
                )
        ) {
            isEditPresented = true
        }
                .sheetEnv(isPresented: $isEditPresented) {
                    FolderFormSheet(
                            isPresented: $isEditPresented,
                            folder: folder
                    )
                }
    }
}
